src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -490,18 +493,38 @@ public synchronized CompletableFuture<SubscribeResponse> subscribe(Subscribe req
      * @throws ExecutionException   if an error occurs
      * @throws InterruptedException if the thread is interrupted while subscribing
      * @throws TimeoutException     if the request times out
+     * @throws MqttException        if the request fails
+     * @deprecated Use {@code subscribe(Subscribe request)} instead
      */
+    @Deprecated
     public void subscribe(SubscribeRequest request)
             throws ExecutionException, InterruptedException, TimeoutException {
         try {
-            Consumer<Publish> cb = (Publish m) -> request.getCallback()
-                    .accept(new MqttMessage(m.getTopic(), m.getPayload(),
-                            QualityOfService.getEnumValueFromInteger(m.getQos().getValue()), m.isRetain()));
-            Subscribe newReq =
-                    Subscribe.builder().qos(QOS.fromInt(request.getQos().getValue())).topic(request.getTopic())
-                            .callback(cb).build();
+            // Deduplicate subscription callbacks so that retries do not result in getting called multiple times
+            Subscribe newReq = cbMapping.computeIfAbsent(new Pair<>(request.getTopic(), request.getCallback()), (p) -> {
+                Consumer<Publish> cb = (Publish m) -> request.getCallback()
+                        .accept(new MqttMessage(m.getTopic(), m.getPayload(),
+                                QualityOfService.getEnumValueFromInteger(m.getQos().getValue()), m.isRetain()));
+
+                return Subscribe.builder().qos(QOS.fromInt(request.getQos().getValue())).topic(request.getTopic())
+                        .callback(cb).build();
+            });
+
             subscribe(newReq)
-                    .thenAccept((v) -> cbMapping.put(new Pair<>(request.getTopic(), request.getCallback()), newReq))
+                    .thenApply((v) -> {
+                        // null is a success because subscribe returns null if the subscription already existed
+                        if (v == null || v.isSuccessful()) {
+                            return v;
+                        }
+                        String rcString = SubAckPacket.SubAckReasonCode.UNSPECIFIED_ERROR.name();
+                        try {
+                            rcString = SubAckPacket.SubAckReasonCode.getEnumValueFromInteger(v.getReasonCode()).name();
+                        } catch (RuntimeException ignored) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception has been ignored. The catch-block hides information about the stack trace. Not logging or forwarding the stack trace of an exception can mask unexpected errors and complicated debugging.

**Fix**: Use a catch block to handle the exception and write a log message declaring that the exception happened. 
[Learn more.](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Suggested remediation:
Handle all exceptions by creating a log for each exception using the message thrown.
```
@@ -522,2 +522,3 @@
                         } catch (RuntimeException ignored) {
+                            log.error(&#34;RuntimeException is caught.&#34;, ignored);
                         }

```